File: Roman Numerals Helper/C/solution_1.c
#include <string.h>

struct RomanMap {
    char* r;
    unsigned int n;
};

const struct RomanMap roman_map[13] = {
    {"M", 1000}, {"CM", 900}, {"D", 500}, {"CD", 400},
    {"C", 100}, {"XC", 90}, {"L", 50}, {"XL", 40},
    {"X", 10}, {"IX", 9}, {"V", 5}, {"IV", 4}, {"I", 1}
};

void to_roman(unsigned int num, char* roman) {
    *roman = '\0';
    while (num > 0) {
        for (int i = 0; i < 13; i++) {
            while (num >= roman_map[i].n) {
                strcat(roman, roman_map[i].r);
                num -= roman_map[i].n;
            }
        }
    }
}

unsigned int from_roman(const char* roman_num) {
    unsigned int num = 0;
    while (*roman_num != '\0') {
        for (int i = 0; i < 13; i++) {
            int r_len = strlen(roman_map[i].r);
            while (strncmp(roman_num, roman_map[i].r, r_len) == 0) {
                num += roman_map[i].n;
                roman_num += r_len;
            }
        }
    }
    return num;
}

File: Roman Numerals Helper/Scala/solution_1.scala
class RomanNumerals {
    def fromRoman(romanNumeral: String): Int = {
        var D = Array(1000, 500, 100, 50, 10, 5, 1)
        var result = 0
        var last = 0
        for (c <- romanNumeral) {
            val index = "MDCLXVI".indexOf(c)
            val current = D(index)
            result += current
            if (current > last) {
                result -= 2 * last
            }
            last = current
        }
        result
    }
    
    def toRoman(n: Int): String = {
        var roman = ""
        var num = n
        val d = Array("M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I")
        val v = Array(1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1)
        for (i <- 0 until d.length) {
            while (num >= v(i)) {
                roman += d(i)
                num -= v(i)
            }
        }
        roman
    }
}